primitives/arithmetic/src/fixed128.rs

```diff
@@ -0,0 +1,351 @@
+};
+use codec::{Encode, Decode};
+use crate::{
+	Perquintill,
+	traits::{
+		SaturatedConversion, CheckedSub, CheckedAdd, CheckedDiv, Bounded, UniqueSaturatedInto, Saturating
+	},
+	helpers_128bit::multiply_by_rational,
+};
+
+/// A signed fixed point number. Can hold any value in the range
+/// [-170_141_183_460_469_231_731, 170_141_183_460_469_231_731]
+/// with an additional fixed point accuracy of 18 decimal points.
+/// i.e. 170000000000000000000.000000000000000001 can be represented by Fixed128.
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed128(i128);
+
+/// The accuracy of the `Fixed128` type.
+const DIV: i128 = 1_000_000_000_000_000_000;
+const U_DIV: u128 = 1_000_000_000_000_000_000;
+
+impl Fixed128 {
+	/// creates self from a natural number.
+	///
+	/// Note that this might be lossy.
+	pub fn from_natural(int: i128) -> Self {
+		Self(int.saturating_mul(DIV))
+	}
+
+	/// Return the accuracy of the type. Given that this function returns the value `X`, it means
+	/// that an instance composed of `X` parts (`Fixed128::from_parts(X)`) is equal to `1`.
+	pub fn accuracy() -> i128 {
+		DIV
+	}
+
+	/// Consume self and return the inner value.
+	///
+	/// This should only be used for testing.
+	#[cfg(any(feature = "std", test))]
+	pub fn into_inner(self) -> i128 { self.0 }
+
+	/// Raw constructor. Equal to `parts / 1_000_000_000_000_000_000`.
+	pub fn from_parts(parts: i128) -> Self {
+		Self(parts)
+	}
+
+	/// creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational(n: i128, d: u128) -> Self {
+		// Turn n into a positive number, and track the original sign.
+		// This lossy by 1 if n is exactly i128.min_value()
+		let (n, sign): (u128, i128) = if n == i128::min_value() {
+			(i128::max_value().try_into().expect("i128 max value can fit into u128"), -1)
+		} else if n.is_negative() {
+			((n * -1).try_into().expect("positive i128 can always fit into u128"), -1)
+		} else {
+			(n.try_into().expect("positive i128 can always fit into u128"), 1)
+		};
+		// This should never fail, but if it does, it is because value was too big for u128
+		// so we map to u128 max value. TODO: Can we expect here that value is always smaller?
+		let result = multiply_by_rational(n, U_DIV, d.max(1)).unwrap_or(u128::max_value());
+		// If the u128 result wont fit in i128, we do i128 max. TODO: Can we expect value is always smaller?
+		let signed_result: i128 = sign * result.try_into().unwrap_or(i128::max_value());
+		Self(signed_result)
+	}
+
+	/// Performs a saturated multiply and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	pub fn saturated_multiply_accumulate<N>(self, int: N) -> N
+		where
+			N: TryFrom<u128> + From<u64> + UniqueSaturatedInto<u64> + Bounded + Clone + Saturating +
+			ops::Rem<N, Output=N> + ops::Div<N, Output=N> + ops::Mul<N, Output=N> +
+			ops::Add<N, Output=N>,
+	{
```

Comment:
These trait bounds are wrong I think for use in the runtime

Reply:
I think we will have to live with it, and just use saturating math within the runtime where needed.